### PR TITLE
Implement data-loss prevention measures for quiz responses

### PR DESCRIPTION
**Pull request by OPS-PIvers**
This pull request introduces robust protections against data loss and regression in student quiz responses, focusing on both client-side and server-side safeguards. The main changes add a per-question append-only answer history for recovery, implement defensive checks to prevent overwriting or downgrading answers due to race conditions, and ensure proper cleanup of historical data. Additional tests validate the new in-memory caching logic that prevents lost answers during navigation and snapshot races.

**Recovery and Data Integrity Improvements**

* Added a Firestore subcollection (`history`) for each student response to store an append-only log of previous answers before they are overwritten, enabling recovery from lost or overwritten student work. Strict security rules restrict creation and reading to the owning student, teachers, or admins, and enforce a fixed schema and server timestamps.
* Implemented logic in `useQuizSession.ts` to throttle history snapshots per question (default: 5 seconds), preventing excessive writes while ensuring meaningful recovery points. [[1]](diffhunk://#diff-c3df9b2601aa95b7241a4e7891e3e652338904d1cf5594edded74f51a30ee13dR68-R86) [[2]](diffhunk://#diff-c3df9b2601aa95b7241a4e7891e3e652338904d1cf5594edded74f51a30ee13dR1247-R1250)

**Client-side Data Loss Prevention**

* Introduced two defensive predicates: `isUnsafeBlankDraft` to block draft autosaves that would overwrite a non-empty answer with a blank (avoiding accidental erasure from race conditions), and `isUnsafeStatusDowngrade` to prevent autosaves from downgrading a submitted answer to draft status (avoiding loss of grading state). [[1]](diffhunk://#diff-c3df9b2601aa95b7241a4e7891e3e652338904d1cf5594edded74f51a30ee13dR112-R170) [[2]](diffhunk://#diff-c3df9b2601aa95b7241a4e7891e3e652338904d1cf5594edded74f51a30ee13dR2036-R2052)
* Updated the student answer submission flow to snapshot the prior answer to history (if eligible) before overwriting, and to apply the above guards on every save. [[1]](diffhunk://#diff-c3df9b2601aa95b7241a4e7891e3e652338904d1cf5594edded74f51a30ee13dR2036-R2052) [[2]](diffhunk://#diff-c3df9b2601aa95b7241a4e7891e3e652338904d1cf5594edded74f51a30ee13dR2102-R2157)

**Cleanup and Maintenance**

* Ensured that when a teacher deletes a student response, all associated answer history documents are deleted first to prevent orphaned recovery data and potential privacy leaks if a student rejoins.

**Testing and Validation**

* Added new tests to verify that the in-memory answer cache preserves student work across navigation and snapshot races, addressing scenarios where answers could be lost due to delayed Firestore updates or hydration issues.

These changes collectively make student answer handling more robust against both user error and technical edge cases, while providing tools for recovery in the event of unexpected data loss.…history log (#1741)

* fix(quiz): prevent student-text loss via cache + blank-write guard + history log

Three layered fixes for the data-loss patterns reported on the live "Independent Reading Test" essay quiz (students seeing previously-typed answers come back blank after a teacher unlock, and back-navigation losing text on previously-submitted questions).

1. Defensive autosave guard (`isUnsafeBlankDraft`). Refuses a draft autosave that would replace a non-empty saved answer with the empty string — almost always a race symptom rather than a deliberate clear. Explicit submits still pass through; only the debounced background path is gated.

2. In-memory per-question answer cache. Replaces the per-type live state slots (`writtenAnswer`, `fibAnswer`, `draftMcAnswer`, the structured ref) with a single `Map<questionId, string>` that's the source of truth for editors and inputs. Navigation reads from the cache instead of rehydrating from the Firestore snapshot on every question change, so the snapshot-listener race that produced blank-after-resume and blank-after-back-nav can no longer drop the student's local edits. `scheduleDraftAutosave` collapses into the existing state-driven autosave effect since matching/ordering now flow through the cache.

3. Append-only `responses/{rid}/history` subcollection. Each prior non-empty answer is snapshotted before being overwritten, throttled per-question to ≥5s so a long essay can't fan out to hundreds of docs. Pure safety net for anything the other two guards miss; teacher / admin reads the subcollection on demand to recover text that a future bug clobbers. Firestore rules pin field shape and immutability.

Tests cover the two pure predicates directly (10 cases across the guard + history-throttle decision) and add two component-level cache behavior tests (navigation preserves text without a Firestore round trip; a late-arriving empty snapshot doesn't clobber a local selection).

* fix(quiz): address code-review findings on PR 1741

Follow-up to the data-loss prevention PR after a high-effort code review surfaced several correctness gaps. All fixes are surgical layered on the cache + history + guard architecture from the parent commit.

- New `isUnsafeStatusDowngrade` predicate refuses a draft autosave that would silently flip an already-submitted answer back to 'draft'. The back-nav listener-lag race in self-paced mode could otherwise rewrite a 'submitted' entry with status='draft' when the response listener hadn't yet echoed the just-submitted answer, dropping the question from the teacher's "Finished" view.

- `shouldSnapshotHistory` now captures snapshots on status downgrades even when the answer text is unchanged. Without this, the same race above destroyed the 'submitted' fact with no recovery record.

- History throttle slot is consumed only AFTER `addDoc` resolves. A failed snapshot (offline, transient permission-denied) used to suppress the next legitimate snapshot for the throttle window even though no recovery doc had been written.

- `lastHistorySnapshotAtRef` cleared on every `joinQuizSession` so a teacher-unlock / attempt-cap rejoin doesn't inherit stale throttle entries from the prior attempt.

- MC locked-branch highlight falls back to the cache value when `selectedAnswer` is null (timer auto-submit never sets it), so an auto-submitted MC question shows which option was committed instead of rendering no highlight at all.

- Reveal-feedback effect consults the cache before falling back to `myResponse.answers` — fixes false-incorrect feedback after timer auto-submit when the Firestore echo lags the teacher's reveal tick.

- WrittenResponseEditor's `questionKey` now encodes whether the cache has been seeded for the current question. A page refresh that mounts the editor before the Firestore snapshot arrives now triggers a one-time remount with the recovered text when the cache populates, instead of leaving the editor blank with the recovered value trapped in React state.

- The autosave effect locally detects the cleared-saved-answer case (cache='', saved=text) and surfaces a `SaveErrorBanner` instructing the student to use Submit to persist the clear — closes the silent trap where the hook's `isUnsafeBlankDraft` refused the write with no UI signal.

Tests: added 4 cases for `isUnsafeStatusDowngrade` and 1 for the status-downgrade snapshot path. Existing 102 quiz tests still pass.

* fix(quiz): cascade-delete answer history + render-body ref sync

Addresses Gemini code-review feedback on PR #1741.

- removeStudent now sweeps the responses/{key}/history subcollection (chunked at 450/batch) before deleting the parent response. Firestore doesn't cascade subcollection deletes, so these snapshots orphaned — and because the response key is deterministic, a removed-then-rejoined student could read their own prior-attempt drafts. The teacher archive still preserves final answers; intermediate-draft history is discarded on removal.

- History snapshot write narrows status to ('draft' ? 'draft' : 'submitted') so a future schema value can't silently fail the rule, and the fire-and-forget catch now logs instead of swallowing — a silently-broken recovery net should be visible.

- Moved the currentQuestion/selectedAnswer/onAnswer/myResponseStatus ref sync from a useEffect into direct render-body assignment, matching answerCacheRef/submittedRef and the project's ref-sync convention. Every consumer reads from an effect or event handler that runs after commit, so this is strictly fresher with no staleness window.

---------

### CODE REVIEW

**Comment by OPS-PIvers** [comment]
Added changelog entry — 4 details, 2 overview themes. Review the copy before merge.

Commit: c57412060b5b26e8803cdb1ca0961ce0f26114eb

Theme: **"Live quiz answer integrity safeguards"**

The recovery-history log (`responses/{rid}/history` subcollection) is **not** mentioned in the changelog because no teacher-facing recovery UI ships in this PR — it's a silent safety net. Happy to add an entry the day the recovery UI lands.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: components/quiz/QuizStudentApp.tsx [hunk 1]
```diff
@@ -2115,10 +2039,10 @@ const ActiveQuiz: React.FC<{
                     {saveError && <SaveErrorBanner message={saveError} />}
                     <button
                       onClick={() =>
-                        fibAnswer.trim() &&
-                        void handleSubmitAndAdvance(fibAnswer.trim())
+                        (liveAnswer ?? '').trim() &&
+                        void handleSubmitAndAdvance((liveAnswer ?? '').trim())
                       }
-                      disabled={!fibAnswer.trim() || submitting}
+                      disabled={!(liveAnswer ?? '').trim() || submitting}
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/quiz/QuizStudentApp.tsx]: Valid — there's a real UX deadlock. The autosave's deliberate-clear branch surfaces "Click Submit to clear it" when a student empties a previously-saved FIB, but this button (along with the teacher-paced one and the Enter-key handler) gated on `(liveAnswer ?? '').trim()` — so the student had no way out of the banner.

Fix pushed in #1745: all three FIB submit affordances now accept an empty submit when `savedAnswerForCurrent` is set. I went with `savedAnswerForCurrent` rather than just falsy-disable-removal so a fresh question without any saved value still can't submit empty (matches the pre-clear-path intent).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: components/quiz/QuizStudentApp.tsx [hunk 2]
```diff
@@ -2140,9 +2064,10 @@ const ActiveQuiz: React.FC<{
               ) : !submitted ? (
                 <button
                   onClick={() =>
-                    fibAnswer.trim() && void handleSubmit(fibAnswer.trim())
+                    (liveAnswer ?? '').trim() &&
+                    void handleSubmit((liveAnswer ?? '').trim())
                   }
-                  disabled={!fibAnswer.trim() || submitting}
+                  disabled={!(liveAnswer ?? '').trim() || submitting}
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/quiz/QuizStudentApp.tsx]: Valid — same deadlock as the self-paced button above. Fix pushed in #1745: this teacher-paced Submit Answer button now also accepts empty submits when `savedAnswerForCurrent` is set, alongside the Enter-key handler.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: hooks/useQuizSession.ts [hunk 3]
```diff
@@ -1962,6 +2099,62 @@ export const useQuizSessionStudent = (): UseQuizSessionStudentResult => {
...
+            QUIZ_SESSIONS_COLLECTION,
+            sessionId,
+            RESPONSES_COLLECTION,
+            responseKey,
+            RESPONSE_HISTORY_COLLECTION
+          ),
+          {
+            questionId: priorEntry.questionId,
+            answer: priorEntry.answer,
+            answeredAt: priorEntry.answeredAt,
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply hooks/useQuizSession.ts]: Valid (low-likelihood but cheap). `answeredAt` is typed `number` (required) and every write path stamps `Date.now()`, so any doc this codebase has ever written has it — but legacy docs predating the field (or any future schema drift) would crash `addDoc` with "Unsupported field value: undefined" and silently break the recovery net.

Fix pushed in #1745: `answeredAt: priorEntry.answeredAt ?? now`. Falling back to `now` (rather than `0`) keeps the snapshot's timeline coherent if it's ever queried.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: hooks/useQuizSession.ts [hunk 4]
```diff
@@ -89,6 +109,65 @@ export function quizLedgerKey(quizId: string, studentUid: string): string {
...
+  lastSnapshotAt: number,
+  now: number,
+  throttleMs: number = HISTORY_SNAPSHOT_THROTTLE_MS
+): boolean {
+  if (!priorEntry) return false;
+  if (priorEntry.answer === '') return false;
+  const textChanged = priorEntry.answer !== newAnswer;
+  const statusDowngrade = priorEntry.status === 'submitted' && newIsDraft;
+  if (!textChanged && !statusDowngrade) return false;
+  return now - lastSnapshotAt >= throttleMs;
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply hooks/useQuizSession.ts:175]: Valid — `submitAnswer` returns early on `isUnsafeStatusDowngrade` BEFORE `shouldSnapshotHistory` runs, so the downgrade branch was dead code in production. The intent (per the commit message) was to record a forensic snapshot of the downgrade attempt for recovery; rejecting the write at the guard meant we silently lost that signal.

Fix pushed in #1745: moved the snapshot decision to BEFORE the downgrade guard. A rejected downgrade now still writes the recovery doc, so the snapshot path matches the comment's stated intent and the existing test exercises a reachable code path. (Updated the related test in #1745 too — see the follow-up reply on the test thread.)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: hooks/useQuizSession.ts [hunk 5]
```diff
@@ -1962,6 +2099,62 @@ export const useQuizSessionStudent = (): UseQuizSessionStudentResult => {
...
+            snapshotAt: serverTimestamp(),
+          }
+        )
+          .then(() => {
+            // Only consume the throttle slot for snapshots that actually
+            // landed. A failed write (offline, rule denial, quota) used
+            // to bump `now` regardless, suppressing the next legitimate
+            // snapshot for the throttle window even though no recovery
+            // doc had been written.
+            lastHistorySnapshotAtRef.current.set(questionId, now);
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply hooks/useQuizSession.ts]: Valid on both counts (out-of-order resolve AND multi-in-flight inside the throttle window).

Fix pushed in #1745: bump `lastHistorySnapshotAtRef` eagerly BEFORE issuing `addDoc`, not in `.then()`. This collapses both concerns — no in-flight writes can see a stale `lastAt`, and out-of-order resolution can no longer move the timestamp backwards (no `.then()` writer at all). The original comment defended `.then()` as "don't burn the slot on failure," but for a fire-and-forget safety net I'd rather take one extra recovery doc in the next window than allow unbounded write-amplification during a network blip. Logged the `.catch()` so silent rules/schema drift remains visible.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/hooks/useQuizSession.test.ts [hunk 6]
```diff
@@ -469,6 +473,183 @@ describe('toPublicQuestion', () => {
...
+    expect(
+      shouldSnapshotHistory(
+        priorSubmitted,
+        'committed answer',
+        true,
+        0,
+        1e9,
+        THROTTLE
+      )
+    ).toBe(true);
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply tests/hooks/useQuizSession.test.ts:658]: Same root cause as the `shouldSnapshotHistory` thread. Rather than retire the test (which characterises a real-world race — back-nav listener-lag silently downgrading a submitted answer), the fix in #1745 makes the scenario reachable by moving the snapshot decision in `submitAnswer` to run BEFORE `isUnsafeStatusDowngrade`. The test now exercises a live code path; the rejected downgrade attempt is recorded as a forensic recovery doc.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: components/quiz/QuizStudentApp.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/quiz/QuizStudentApp.tsx]: Addressed in #1744 (merged): the FIB/essay submit now gates on the cache-backed `submittableAnswer`, and a deliberate clear holds a non-null empty string, so the button stays enabled and the clear can be submitted. Consolidated onto dev-paul.

File: components/quiz/QuizStudentApp.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/quiz/QuizStudentApp.tsx]: Addressed in #1744 (merged): teacher-paced Submit gates on `submittableAnswer` as well; a deliberate clear is a non-null empty string, so Submit stays enabled. Consolidated onto dev-paul.

File: hooks/useQuizSession.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply hooks/useQuizSession.ts]: Addressed in #1745 (merged): the history snapshot now writes `answeredAt: priorEntry.answeredAt ?? now`, so a legacy or missing timestamp can no longer pass undefined to addDoc.

File: components/quiz/QuizStudentApp.tsx [hunk 7]
```diff
@@ -2181,14 +2106,13 @@ const ActiveQuiz: React.FC<{
             question={currentQuestion}
             submitted={submitted}
             isAutoSubmitted={autoSubmitTriggeredFor === currentQuestion.id}
-            savedAnswer={savedAnswerForCurrent}
+            savedAnswer={liveAnswer}
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/quiz/QuizStudentApp.tsx:2212]: Addressed on dev-paul (6757cb15): untouched questions stay seed-tracked via a refresh-untouched seed-from-server block, and the StructuredQuestionInput mount re-emit is guarded, so a late snapshot updates the cache instead of being clobbered. Regression test added.

File: hooks/useQuizSession.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply hooks/useQuizSession.ts:175]: Addressed in #1745 (merged): the history-snapshot decision now runs BEFORE the `isUnsafeStatusDowngrade` guard, so the downgrade path is reachable and a rejected downgrade attempt is still recorded.

File: hooks/useQuizSession.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply hooks/useQuizSession.ts]: Addressed in #1745 (merged): the throttle slot is now consumed eagerly (before addDoc), so concurrent in-flight writes cannot all read a stale lastAt and slip past the throttle.

File: tests/hooks/useQuizSession.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply tests/hooks/useQuizSession.test.ts:658]: Addressed in #1745 (merged): the snapshot decision now runs before the downgrade guard, so the downgrade-snapshot scenario this test covers is reachable in production.
